README.md
# ✔️ ZEST: Zestful Testing ✔️

---

## JS Testing Stack + Auto-Runner + Live Terminal Reporting Interface

### [Mocha + Chai + SuperTest] + [Blessed + Blessed-Contrib + Blessed-XTerm]

---

<pre>
    ✔️ Zest can run plain mocha tests!
    ✔️ Mocha can run fancy zest tests!
</pre>


CHANGELOG.md
<code>

# [0.0.29](https://github.com/cogsmith/zest/compare/0.0.29...main) @ [2021-08-30](https://github.com/cogsmith/zest/releases/tag/0.0.29) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.28](https://github.com/cogsmith/zest/compare/0.0.28...0.0.29)

</code>

---
<code>

# [0.0.28](https://github.com/cogsmith/zest/compare/0.0.28...main) @ [2021-08-30](https://github.com/cogsmith/zest/releases/tag/0.0.28) 

## DIFF
- [3 COMMITS SINCE LAST TAG = 0.0.27](https://github.com/cogsmith/zest/compare/0.0.27...0.0.28)

</code>

---
<code>

# [0.0.27](https://github.com/cogsmith/zest/compare/0.0.27...main) @ [2021-08-29](https://github.com/cogsmith/zest/releases/tag/0.0.27) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.26](https://github.com/cogsmith/zest/compare/0.0.26...0.0.27)

</code>

---
<code>

# [0.0.26](https://github.com/cogsmith/zest/compare/0.0.26...main) @ [2021-08-29](https://github.com/cogsmith/zest/releases/tag/0.0.26) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.25](https://github.com/cogsmith/zest/compare/0.0.25...0.0.26)

</code>

---
<code>

# [0.0.25](https://github.com/cogsmith/zest/compare/0.0.25...main) @ [2021-08-29](https://github.com/cogsmith/zest/releases/tag/0.0.25) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.24](https://github.com/cogsmith/zest/compare/0.0.24...0.0.25)

</code>

---
<code>

# [0.0.24](https://github.com/cogsmith/zest/compare/0.0.24...main) @ [2021-08-29](https://github.com/cogsmith/zest/releases/tag/0.0.24) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.23](https://github.com/cogsmith/zest/compare/0.0.23...0.0.24)

</code>

---
<code>

# [0.0.23](https://github.com/cogsmith/zest/compare/0.0.23...main) @ [2021-08-29](https://github.com/cogsmith/zest/releases/tag/0.0.23) 

## DIFF
- [5 COMMITS SINCE LAST TAG = 0.0.22](https://github.com/cogsmith/zest/compare/0.0.22...0.0.23)

</code>

---
<code>

# [0.0.22](https://github.com/cogsmith/zest/compare/0.0.22...main) @ [2021-08-12](https://github.com/cogsmith/zest/releases/tag/0.0.22) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.21](https://github.com/cogsmith/zest/compare/0.0.21...0.0.22)

</code>

---
<code>

# [0.0.21](https://github.com/cogsmith/zest/compare/0.0.21...main) @ [2021-08-12](https://github.com/cogsmith/zest/releases/tag/0.0.21) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.20](https://github.com/cogsmith/zest/compare/0.0.20...0.0.21)

</code>

---
<code>

# [0.0.20](https://github.com/cogsmith/zest/compare/0.0.20...main) @ [2021-08-12](https://github.com/cogsmith/zest/releases/tag/0.0.20) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.19](https://github.com/cogsmith/zest/compare/0.0.19...0.0.20)

</code>

---
<code>

# [0.0.19](https://github.com/cogsmith/zest/compare/0.0.19...main) @ [2021-08-08](https://github.com/cogsmith/zest/releases/tag/0.0.19) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.18](https://github.com/cogsmith/zest/compare/0.0.18...0.0.19)

</code>

---
<code>

# [0.0.18](https://github.com/cogsmith/zest/compare/0.0.18...main) @ [2021-08-08](https://github.com/cogsmith/zest/releases/tag/0.0.18) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.17](https://github.com/cogsmith/zest/compare/0.0.17...0.0.18)

</code>

---
<code>

# [0.0.17](https://github.com/cogsmith/zest/compare/0.0.17...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.17) 

## DIFF
- [3 COMMITS SINCE LAST TAG = 0.0.16](https://github.com/cogsmith/zest/compare/0.0.16...0.0.17)

</code>

---
<code>

# [0.0.16](https://github.com/cogsmith/zest/compare/0.0.16...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.16) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.15](https://github.com/cogsmith/zest/compare/0.0.15...0.0.16)

</code>

---
<code>

# [0.0.15](https://github.com/cogsmith/zest/compare/0.0.15...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.15) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.14](https://github.com/cogsmith/zest/compare/0.0.14...0.0.15)

</code>

---
<code>

# [0.0.14](https://github.com/cogsmith/zest/compare/0.0.14...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.14) 

## DIFF
- [4 COMMITS SINCE LAST TAG = 0.0.13](https://github.com/cogsmith/zest/compare/0.0.13...0.0.14)

## COMMITS
- Merge branch 'main' of https://github.com/cogsmith/zest

</code>

---
<code>

# [0.0.13](https://github.com/cogsmith/zest/compare/0.0.13...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.13) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.12](https://github.com/cogsmith/zest/compare/0.0.12...0.0.13)

</code>

---
<code>

# [0.0.12](https://github.com/cogsmith/zest/compare/0.0.12...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.12) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.11](https://github.com/cogsmith/zest/compare/0.0.11...0.0.12)

</code>

---
<code>

# [0.0.11](https://github.com/cogsmith/zest/compare/0.0.11...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.11) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.10](https://github.com/cogsmith/zest/compare/0.0.10...0.0.11)

</code>

---
<code>

# [0.0.10](https://github.com/cogsmith/zest/compare/0.0.10...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.10) 

## DIFF
- [4 COMMITS SINCE LAST TAG = 0.0.9](https://github.com/cogsmith/zest/compare/0.0.9...0.0.10)

</code>

---
<code>

# [0.0.9](https://github.com/cogsmith/zest/compare/0.0.9...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.9) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.8](https://github.com/cogsmith/zest/compare/0.0.8...0.0.9)

</code>

---
<code>

# [0.0.8](https://github.com/cogsmith/zest/compare/0.0.8...main) @ [2021-08-07](https://github.com/cogsmith/zest/releases/tag/0.0.8) 

## DIFF
- [2 COMMITS SINCE LAST TAG = 0.0.7](https://github.com/cogsmith/zest/compare/0.0.7...0.0.8)

</code>

---
<code>


test/supertest.js
const assert = require('assert');
const supertest = require('supertest');
let ff = 0;

const Mock = {};
Mock.Host = process.env.HOST || 'localhost';
Mock.Port = process.env.PORT || 8080;
Mock.URL = 'http://' + Mock.Host.trim() + ':' + Mock.Port.toString().trim();
Mock.Init = async () => { await Mock.Fixture(); };
Mock.Done = async () => { await ff.close(); };
Mock.Reset = async () => { await Mock.Done(); await Mock.Init(); };
Mock.Fixture = async () => {
    ff = require('fastify')({ logger: 0 });
    ff.get('/', async (req, rep) => { return 'HELLOWORLD'; });
    ff.get('/404', async (req, rep) => { rep.status(404).send('404'); });
    ff.get('/status/200', async (req, rep) => { rep.status(200).send('200'); });
    ff.get('/status/400', async (req, rep) => { rep.status(400).send('400'); });
    ff.post('/post', async (req, rep) => { rep.status(200).send('POST'); });
    await ff.listen(Mock.Port);
};

const web = supertest(Mock.URL);
describe('Test Suite', () => {
    before(Mock.Init);
    after(Mock.Done);
    afterEach(Mock.Reset);
    //beforeEach(Mock.Reset);
    describe('Status Codes', () => {
        it('200', async () => { await web.get('/status/200').expect(200); });
        it('400', async () => { await web.get('/status/400').expect(400); });
        it('404', async () => { await web.get('/404').expect(404); });
        it('TESTFAIL', async () => { await web.get('/').expect(9); });
    });
    describe('Response Body', () => {
        it('HELLOWORLD', async () => { await web.get('/').expect('HELLOWORLD'); });
        it('POST', async () => { await web.post('/post').expect(200); });
    });
});

// CMD /C "SET HOST=google.com & SET PORT=80 & npx -y @cogsmith/zest"
// CMD /C "SET HOST=localhost & SET PORT=8080 & npx -y @cogsmith/zest"


app.js
#!/usr/bin/env node

const XT = require('@cogsmith/xt').Init();
const App = XT.App; const LOG = XT.LOG;
//XT.Log.SetLevel('warn');

//

const Zest = require('./lib');

//

const NODE = {};
NODE.fs = require('fs');
NODE.http = require('http');

const LIB = {};
LIB.mocha = require('mocha'); const mocha = LIB.mocha;
LIB.chalk = require('chalk'); const chalk = LIB.chalk;

//

App.InitLog = function () {
    //XT.Log.SetLevel('warn');
}

App.InitArgs = function () {
    // App.Argy = yargs(process.argv);
}

App.InitInfo = function () {
    // App.SetInfo('App', function () { return 'ZEST' });
}

App.InitData = function () {
}

App.Init = function () {

}

App.InitDone = function () {
}

App.Main = async function () {
    //LOG.WARN('App.Args', App.Args);
    //LOG.WARN('', { PCWD: process.cwd(), DIRNAME: __dirname });
    //LOG.WARN(XT.Meta.Full);
    //LOG.WARN(App.Meta.Full);

    setTimeout(async () => { await Zest.AppMain(App); }, 2500);
}

App.Run();


reporters/base.js
// MOCHA REPORTER BASE + TWEAKS

const util = require('util');
const mocha = require('mocha'); const { EVENT_RUN_BEGIN, EVENT_RUN_END, EVENT_TEST_FAIL, EVENT_TEST_PASS, EVENT_SUITE_BEGIN, EVENT_SUITE_END, EVENT_TEST_PENDING } = mocha.Runner.constants;

let RUNNER = false;

'use strict';
/**
 * @module Base
 */
/**
 * Module dependencies.
 */

var diff = require('diff');
var milliseconds = require('ms');
//var utils = require('../utils');
var supportsColor = require('supports-color');
var symbols = require('log-symbols');
//var constants = require('../runner').constants;
//var EVENT_TEST_PASS = constants.EVENT_TEST_PASS;
//var EVENT_TEST_FAIL = constants.EVENT_TEST_FAIL;

// const isBrowser = utils.isBrowser();
const isBrowser = false;

function getBrowserWindowSize() {
    if ('innerHeight' in global) {
        return [global.innerHeight, global.innerWidth];
    }
    // In a Web Worker, the DOM Window is not available.
    return [640, 480];
}

/**
 * Expose `Base`.
 */

exports = module.exports = Base;

/**
 * Check if both stdio streams are associated with a tty.
 */

var isatty = isBrowser || (process.stdout.isTTY && process.stderr.isTTY);

/**
 * Save log references to avoid tests interfering (see GH-3604).
 */
//var consoleLog = console.log;
/*
let consoleLog = function () {
    //console.log(...arguments);
    //Base.RUNNER.LOGNOTES('FOO' + "\n");
}
*/

/**
 * Enable coloring by default, except in the browser interface.
 */

exports.useColors =
    !isBrowser &&
    (supportsColor.stdout || process.env.MOCHA_COLORS !== undefined);

/**
 * Inline diffs instead of +/-
 */

exports.inlineDiffs = false;

/**
 * Default color map.
 */

exports.colors = {
    pass: 90,
    fail: 31,
    'bright pass': 92,
    'bright fail': 91,
    'bright yellow': 93,
    pending: 36,
    suite: 0,
    'error title': 0,
    'error message': 31,
    'error stack': 90,
    checkmark: 32,
    fast: 90,
    medium: 33,
    slow: 31,
    green: 32,
    light: 90,
    'diff gutter': 90,
    'diff added': 32,
    'diff removed': 31,
    'diff added inline': '30;42',
    'diff removed inline': '30;41'
};

/**
 * Default symbol map.
 */

exports.symbols = {
    ok: symbols.success,
    err: symbols.error,
    dot: '.',
    comma: ',',
    bang: '!'
};

/**
 * Color `str` with the given `type`,
 * allowing colors to be disabled,
 * as well as user-defined color
 * schemes.
 *
 * @private
 * @param {string} type
 * @param {string} str
 * @return {string}
 */
var color = (exports.color = function (type, str) {
    if (!exports.useColors) {
        return String(str);
    }
    return '\u001b[' + exports.colors[type] + 'm' + str + '\u001b[0m';
});

/**
 * Expose term window size, with some defaults for when stderr is not a tty.
 */

exports.window = {
    width: 75
};

if (isatty) {
    if (isBrowser) {
        exports.window.width = getBrowserWindowSize()[1];
    } else {
        exports.window.width = process.stdout.getWindowSize(1)[0];
    }
}

/**
 * Expose some basic cursor interactions that are common among reporters.
 */

exports.cursor = {
    hide: function () {
        isatty && process.stdout.write('\u001b[?25l');
    },

    show: function () {
        isatty && process.stdout.write('\u001b[?25h');
    },

    deleteLine: function () {
        isatty && process.stdout.write('\u001b[2K');
    },

    beginningOfLine: function () {
        isatty && process.stdout.write('\u001b[0G');
    },

    CR: function () {
        if (isatty) {
            exports.cursor.deleteLine();
            exports.cursor.beginningOfLine();
        } else {
            process.stdout.write('\r');
        }
    }
};

var showDiff = (exports.showDiff = function (err) {
    return (
        err &&
        err.showDiff !== false &&
        sameType(err.actual, err.expected) &&
        err.expected !== undefined
    );
});

function stringifyDiffObjs(err) {
    err.actual = err.actual;
    err.expected = err.expected;

    /*
    if (!utils.isString(err.actual) || !utils.isString(err.expected)) {
        err.actual = utils.stringify(err.actual);
        err.expected = utils.stringify(err.expected);
    }
    */
}

/**
 * Returns a diff between 2 strings with coloured ANSI output.
 *
 * @description
 * The diff will be either inline or unified dependent on the value
 * of `Base.inlineDiff`.
 *
 * @param {string} actual
 * @param {string} expected
 * @return {string} Diff
 */
var generateDiff = (exports.generateDiff = function (actual, expected) {
    try {
        const diffSize = 2048;
        if (actual.length > diffSize) {
            actual = actual.substring(0, diffSize) + ' ... Lines skipped';
        }
        if (expected.length > diffSize) {
            expected = expected.substring(0, diffSize) + ' ... Lines skipped';
        }
        return exports.inlineDiffs
            ? inlineDiff(actual, expected)
            : unifiedDiff(actual, expected);
    } catch (err) {
        var msg =
            '\n      ' +
            color('diff added', '+ expected') +
            ' ' +
            color('diff removed', '- actual:  failed to generate Mocha diff') +
            '\n';
        return msg;
    }
});

/**
 * Outputs the given `failures` as a list.
 *
 * @public
 * @memberof Mocha.reporters.Base
 * @variation 1
 * @param {Object[]} failures - Each is Test instance with corresponding
 *     Error property
 */
exports.list = function (runner, failures) {
    var multipleErr, multipleTest;
    let BaseConsoleLog = runner.LOG;
    BaseConsoleLog();
    failures.forEach(function (test, i) {
        // format
        var fmt =
            color('error title', '  %s) %s:\n') +
            color('error message', '     %s') +
            color('error stack', '\n%s\n');

        // msg
        var msg;
        var err;
        if (test.err && test.err.multiple) {
            if (multipleTest !== test) {
                multipleTest = test;
                multipleErr = [test.err].concat(test.err.multiple);
            }
            err = multipleErr.shift();
        } else {
            err = test.err;
        }
        var message;
        if (typeof err.inspect === 'function') {
            message = err.inspect() + '';
        } else if (err.message && typeof err.message.toString === 'function') {
            message = err.message + '';
        } else {
            message = '';
        }
        var stack = err.stack || message;
        var index = message ? stack.indexOf(message) : -1;

        if (index === -1) {
            msg = message;
        } else {
            index += message.length;
            msg = stack.slice(0, index);
            // remove msg from stack
            stack = stack.slice(index + 1);
        }

        // uncaught
        if (err.uncaught) {
            msg = 'Uncaught ' + msg;
        }
        // explicitly show diff
        if (!exports.hideDiff && showDiff(err)) {
            stringifyDiffObjs(err);
            fmt =
                color('error title', '  %s) %s:\n%s') + color('error stack', '\n%s\n');
            var match = message.match(/^([^:]+): expected/);
            msg = '\n      ' + color('error message', match ? match[1] : msg);

            msg += generateDiff(err.actual, err.expected);
        }

        // indent stack trace
        stack = stack.replace(/^/gm, '  ');

        // indented test title
        var testTitle = '';
        test.titlePath().forEach(function (str, index) {
            if (index !== 0) {
                testTitle += '\n     ';
            }
            for (var i = 0; i < index; i++) {
                testTitle += '  ';
            }
            testTitle += str;
        });

        BaseConsoleLog(fmt, i + 1, testTitle, msg, stack);
        BaseConsoleLog();
    });
};

/**
 * Constructs a new `Base` reporter instance.
 *
 * @description
 * All other reporters generally inherit from this reporter.
 *
 * @public
 * @class
 * @memberof Mocha.reporters
 * @param {Runner} runner - Instance triggers reporter actions.
 * @param {Object} [options] - runner options
 */
function Base(runner, options) {
    var failures = (this.failures = []);

    if (!runner) {
        throw new TypeError('Missing runner argument');
    }
    this.options = options || {};
    this.runner = runner;
    this.stats = runner.stats; // assigned so Reporters keep a closer reference

    RUNNER = runner;
    global.RUNNER = runner;

    runner.on(EVENT_TEST_PASS, function (test) {
        if (test.duration > test.slow()) {
            test.speed = 'slow';
        } else if (test.duration > test.slow() / 2) {
            test.speed = 'medium';
        } else {
            test.speed = 'fast';
        }
    });

    runner.on(EVENT_TEST_FAIL, function (test, err) {
        if (showDiff(err)) {
            stringifyDiffObjs(err);
        }
        // more than one error per test
        if (test.err && err instanceof Error) {
            test.err.multiple = (test.err.multiple || []).concat(err);
        } else {
            test.err = err;
        }
        failures.push(test);
    });
}

/**
 * Outputs common epilogue used by many of the bundled reporters.
 *
 * @public
 * @memberof Mocha.reporters
 */
Base.prototype.epilogue = function (runner) {
    var stats = this.stats;
    var fmt;

    BaseConsoleLog = runner.LOG;

    BaseConsoleLog();

    // passes
    fmt =
        color('bright pass', ' ') +
        color('green', ' %d passing') +
        color('light', ' (%s)');

    BaseConsoleLog(fmt, stats.passes || 0, milliseconds(stats.duration));

    // pending
    if (stats.pending) {
        fmt = color('pending', ' ') + color('pending', ' %d pending');

        BaseConsoleLog(fmt, stats.pending);
    }

    // failures
    if (stats.failures) {
        fmt = color('fail', '  %d failing');

        BaseConsoleLog(fmt, stats.failures);

        Base.list(runner, this.failures);
        BaseConsoleLog();
    }

    BaseConsoleLog();
};

/**
 * Pads the given `str` to `len`.
 *
 * @private
 * @param {string} str
 * @param {string} len
 * @return {string}
 */
function pad(str, len) {
    str = String(str);
    return Array(len - str.length + 1).join(' ') + str;
}

/**
 * Returns inline diff between 2 strings with coloured ANSI output.
 *
 * @private
 * @param {String} actual
 * @param {String} expected
 * @return {string} Diff
 */
function inlineDiff(actual, expected) {
    var msg = errorDiff(actual, expected);

    // linenos
    var lines = msg.split('\n');
    if (lines.length > 4) {
        var width = String(lines.length).length;
        msg = lines
            .map(function (str, i) {
                return pad(++i, width) + ' |' + ' ' + str;
            })
            .join('\n');
    }

    // legend
    msg =
        '\n' +
        color('diff removed inline', 'actual') +
        ' ' +
        color('diff added inline', 'expected') +
        '\n\n' +
        msg +
        '\n';

    // indent
    msg = msg.replace(/^/gm, '      ');
    return msg;
}

/**
 * Returns unified diff between two strings with coloured ANSI output.
 *
 * @private
 * @param {String} actual
 * @param {String} expected
 * @return {string} The diff.
 */
function unifiedDiff(actual, expected) {
    var indent = '      ';
    function cleanUp(line) {
        if (line[0] === '+') {
            return indent + colorLines('diff added', line);
        }
        if (line[0] === '-') {
            return indent + colorLines('diff removed', line);
        }
        if (line.match(/@@/)) {
            return '--';
        }
        if (line.match(/\\ No newline/)) {
            return null;
        }
        return indent + line;
    }
    function notBlank(line) {
        return typeof line !== 'undefined' && line !== null;
    }
    var msg = diff.createPatch('string', actual, expected);
    var lines = msg.split('\n').splice(5);
    return (
        '\n      ' +
        colorLines('diff added', '+ expected') +
        ' ' +
        colorLines('diff removed', '- actual') +
        '\n\n' +
        lines
            .map(cleanUp)
            .filter(notBlank)
            .join('\n')
    );
}

/**
 * Returns character diff for `err`.
 *
 * @private
 * @param {String} actual
 * @param {String} expected
 * @return {string} the diff
 */
function errorDiff(actual, expected) {
    return diff
        .diffWordsWithSpace(actual, expected)
        .map(function (str) {
            if (str.added) {
                return colorLines('diff added inline', str.value);
            }
            if (str.removed) {
                return colorLines('diff removed inline', str.value);
            }
            return str.value;
        })
        .join('');
}

/**
 * Colors lines for `str`, using the color `name`.
 *
 * @private
 * @param {string} name
 * @param {string} str
 * @return {string}
 */
function colorLines(name, str) {
    return str
        .split('\n')
        .map(function (str) {
            return color(name, str);
        })
        .join('\n');
}

/**
 * Object#toString reference.
 */
var objToString = Object.prototype.toString;

/**
 * Checks that a / b have the same type.
 *
 * @private
 * @param {Object} a
 * @param {Object} b
 * @return {boolean}
 */
function sameType(a, b) {
    return objToString.call(a) === objToString.call(b);
}

Base.abstract = true;


lib.js
//process.on('uncaughtException', function (err) { console.log("\n"); console.log(err); console.log("\n"); process.exit(1); }); // throw(Error('ERROR'));
//process.on('unhandledRejection', function (err) { console.log("\n"); console.log(err); console.log("\n"); process.exit(1); }); // throw(Error('ERROR'));

//

const XT = require('@cogsmith/xt').Init();
const LOG = XT.LOG;

//

const NODE = {};
NODE.fs = require('fs');
NODE.http = require('http');
NODE.path = require('path'); const nodepath = NODE.path;

const LIBS = {};
LIBS.mocha = require('mocha'); const mocha = LIBS.mocha;
LIBS.chalk = require('chalk'); const chalk = LIBS.chalk;
LIBS.glob = require('glob'); const glob = LIBS.glob;
LIBS.supertest = require('supertest'); const web = LIBS.supertest('');
LIBS.chai = require('chai'); const should = LIBS.chai.should();
LIBS.blessed = require('blessed'); const blessed = LIBS.blessed;
LIBS.blessedx = require('blessed-contrib'); const blessedx = LIBS.blessedx;

//

/*
const {
    EVENT_RUN_BEGIN,
    EVENT_RUN_END,
    EVENT_TEST_FAIL,
    EVENT_TEST_PASS,
    EVENT_SUITE_BEGIN,
    EVENT_SUITE_END,
    EVENT_TEST_PENDING
} = LIBS.mocha.Runner.constants;
*/

//

const Zest = {};

Zest.DB = { Tests: {} }; //, TestsList: [] };

Zest.GetTests = () => { return XT.Object.List(Zest.DB.Tests); };

Zest.Test = function () {
    let self = this;
    let id = false;
    let test = {};

    if (arguments.length == 0 || arguments.length > 2) { return; }
    if (arguments.length == 1) {
        let dat = arguments[0];
        id = Object.keys(dat)[0];
        test = dat[id];
    }
    if (arguments.length == 2) {
        id = arguments[0];
        let dat = arguments[1];
        if (typeof dat == 'function') { test.FX = dat; }
        if (typeof dat == 'object') {
            if (dat.Web || dat.FX) { test = dat; }
            if (!dat.Web) { dat.Web = {}; }
            if (dat.URL) {
                dat.Web.URL = dat.URL; delete dat.URL;
                if (dat.Input) { dat.Web.Input = dat.Input; delete dat.Input; }
                if (dat.Output) { dat.Web.Output = dat.Output; delete dat.Output; }
            }
            test = dat;
        }
    }

    test.ID = id;
    Zest.DB.Tests[id] = test;

    describe(test.ID, function () {
        describe('Init Tests', function () {
            if (test.Web && test.Web.URL) {
                it('Fetch Data', async function () {
                    res = await web.post(test.Web.URL).send(test.Web.Input).expect(200);
                })
            }
        })

        describe('Data Validation', function () {
            if (test.Web && test.Web.Output) {
                it(test.ID, async function () {
                    res.body.should.eql(test.Web.Output);
                });
            }
        })

        describe('Test Cases', function () {
            if (test.FX) {
                it(test.ID + ' FX', async function () {
                    await test.FX.apply(test, [res]);
                });
            }
        })
    })
}

Zest.IsPortOpen = (port) => new Promise((resolve, reject) => {
    let server = NODE.http.createServer((req, res) => { res.writeHead(200); res.end('HI'); });
    server.on('error', () => { resolve(false); server.close(); });
    server.on('listening', () => { resolve(true); server.close(); });
    server.listen(port, '127.0.0.1');
})

Zest.WaitForPort = (port) => new Promise(async (resolve, reject) => {
    let timer = setInterval(async () => {
        if (await App.IsPortOpen(port)) { clearInterval(timer); resolve(); }
    }, 10);
})

Zest.UI = {};

Zest.UI.Render = function () { Zest.UI.Screen.render(); }

Zest.InitUI = function () {
    LOG.WARN('InitUI');
    const screen = blessed.screen({ debug: true, dockBorders: true });
    // screen.key(['q', 'C-c'], function (ch, key) { process.exit(0); });
    screen.key(['q', 'C-c'], function (ch, key) { screen.destroy(); });
    let style = { selected: { fg: 'white', bg: 'red' } };
    //let items = ['[ ALL TESTS ]', 'test/supertest.js', 'test/zest.js'];
    let list = Zest.TestList.map(z => { z = z.replace(process.cwd().replace(/\\/g, '/'), ''); return z.substr(1); });
    let items = ['[ ALL TESTS ]', ...list];
    Zest.Layout = new blessedx.grid({ hideBorders: true, rows: 10, cols: 10, screen: screen });
    const grid = Zest.Layout;
    const head = grid.set(0, 0, 2, 4, blessed.list, { interactive: true, keys: true, style, items });
    const foot = grid.set(2, 0, 8, 4, blessed.box, { content: '' });
    const body = grid.set(0, 4, 10, 6, blessed.box, { style: { bg: '#111111' }, content: '' });
    head.focus();
    head.on('select', (z, i) => {
        if (Zest.TestRunning) { return; }
        foot.content = z.content + "\n";
        let test = i; if (i > 0) { test = z.content; };
        Zest.UI.Head.focus();
        Zest.DoTest(test);
        Zest.UI.Head.focus();
    });
    screen.render();

    Zest.UI.Screen = screen;
    Zest.UI.Head = head;
    Zest.UI.Foot = foot;
    Zest.UI.Body = body;
}


XT.Buffer = function (id) {
    if (!id) { id = 0; }
    console.log(id);
    let buf = XT.Buffer.DB[id];
    if (buf) { buf = buf.substr(0, buf.lenth - 2); }
    return buf;
}

XT.Buffer.DB = {};
XT.Buffer.Get = function (id) { return XT.Buffer(id); }
XT.Buffer.Reset = function (id) { XT.Buffer.DB[id] = ''; };
XT.Buffer.Append = function (id, dat) {
    if (!dat) { dat = id; id = 0; };
    if (!XT.Buffer.DB[id]) { XT.Buffer.DB[id] = ''; }
    XT.Buffer.DB[id] += dat;
}

XT.Stream = {};

XT.Stream.Restore = function (s) {
    s.write = s.writebind;
    //delete s.writebind;
}

XT.Stream.SaveAndMute = function (s) {
    XT.Buffer.Reset(s);
    s.writebind = s.write.bind(s);
    s.write = function (dat, enc, nxt) {
        XT.Buffer.Append(dat);
        //return s.writebind.apply(this, arguments);
    }
}

XT.Stream.SaveAndPipe = function (s) {
    XT.Buffer.Reset();
    s.writebind = s.write.bind(s);
    s.write = function (dat, enc, nxt) {
        XT.Buffer.Append(dat);
        return s.writebind.apply(this, arguments);
    }
}

Zest.TestList = [];
Zest.GetTestList = function (arglist) {
    let list = [];
    for (let t of arglist) {
        if (NODE.fs.existsSync(t)) {
            let dir = process.cwd();
            let js = dir + '/' + t;
            if (!NODE.fs.existsSync(js)) { js = t; }
            if (!NODE.fs.existsSync(js)) { continue; }
            if (NODE.fs.lstatSync(js).isDirectory()) {
                for (let f of glob.sync(js + '/**/*.js')) { list.push(f); }
            } else {
                list.push(js);
            }
        }
    }
    return list;
}

Zest.TestRunning = false;
Zest.DoTest = function (testlist) {
    if (Zest.TestRunning) { return; }; Zest.TestRunning = true;
    if (testlist == 0 || testlist == '0') { testlist = Zest.TestList; }
    //if (testlist == 0 || testlist == '0') { testlist = Zest.GetTestList([process.cwd() + '/' + 'test']); }
    if (!Array.isArray(testlist)) { testlist = [testlist]; }
    if (NODE.fs.lstatSync(testlist[0]).isDirectory()) { testlist = Zest.GetTestList(testlist); }

    //XT.Stream.SaveAndMute(process.stdout);
    //process.stdout.write = function () { };

    let mocha = new LIBS.mocha({ reporter: 'spec', ui: 'bdd', timeout: 10000 });
    mocha.reporter(require('./reporters/full'));

    //console.error(require.cache);

    for (let js of testlist) {
        if (js[0] != '.' && js[0] != '/' && !js.includes(':')) { js = './' + js; }
        //console.log(require.resolve(js.replace('.js', '')));
        try {
            let module = nodepath.join(process.cwd(), js.replace('.js', ''));
            delete require.cache[require.resolve(module)];
        } catch (ex) { console.error(ex); }
        mocha.addFile(js);
    }

    //mocha.addFile('test/test1.js');
    let runner = mocha.run(function (fails) {
        //console.log({ RUN: { FAILS: fails, ARGS: arguments } });

        if (Zest.App.Args.ui) {
            //console.error(runner);
            //XT.Stream.Restore(process.stdout);
            //Zest.UI.Body.content = XT.Buffer();
            Zest.UI.Foot.content = runner.SUMMARY;
            Zest.UI.Body.content = runner.LOGOUT;
            Zest.UI.Render();
        } else {
            console.log(runner.SUMMARY);
            console.log(runner.LOGOUT);
        }

        mocha.unloadFiles();
    });

    Zest.UI.Head.focus();
    setTimeout(() => { Zest.TestRunning = false; Zest.UI.Head.focus(); }, 250);
}


Zest.AppMain = async function (App) {
    Zest.App = App;

    //console.clear();

    //console.log('AppArgs', App.Args);

    //console.clear();
    //LOG.WARN('Z');
    //LOG.WARN('ARGS', App.Args);

    //

    if (!App.Args.test) { App.Args.test = []; } else if (!Array.isArray(App.Args.test)) { App.Args.test = [App.Args.test]; }
    if (App.Args._.length > 2) { App.Args.test = App.Args.test.concat(App.Args._.slice(2)); }
    if (App.Args.test.length == 0) { App.Args.test = ['test']; }

    Zest.TestList = Zest.GetTestList(App.Args.test);

    if (App.Args.ui) { Zest.InitUI(); } else { Zest.DoTest(App.Args.test); }

    //Zest.DoTest();
    //setTimeout(Zest.DoTest, 2500);

    if (0 && App.Args.loop) {
        if (App.Args.waitport) { await Zest.WaitForPort(App.Args.waitport); }
        setTimeout(Zest.AppMain, 1000);
    }
}

//

module.exports = Zest;
